aten/src/ATen/native/TensorShape.cpp

```diff
@@ -1465,4 +1465,30 @@ Tensor& diag_out(Tensor &result, const Tensor& self, int64_t dimension) {
   return result;
 }
 
+Tensor trace(const Tensor& self) {
+  Tensor result = at::empty({}, self.options());
+  AT_DISPATCH_ALL_TYPES(self.scalar_type(), "trace", [&] {
+    scalar_t sum = 0;
```

Comment:
We should be performing the sum in an accumulated type, not the scalar type of `self`, to avoid precision issues. Inside the deleted TH code, we can see that it is accumulating into 'accreal'.

So this code should be:
```
// Pseudocode, possibly syntax issues
using accscalar_t = acctype<scalar_t, /*is_cuda*/True>;
```

Here's where acc_type is defined: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/AccumulateType.h

Reply:
done